internal/devpkg/package.go

```diff
@@ -462,37 +499,113 @@ func (p *Package) HashFromNixPkgsURL() string {
 // It is used as FromStore in builtins.fetchClosure.
 const BinaryCache = "https://cache.nixos.org"
 
-func (p *Package) IsInBinaryCache() (bool, error) {
+func (p *Package) isEligibleForBinaryCache() (bool, error) {
+	sysInfo, err := p.sysInfoIfExists()
+	if err != nil {
+		return false, err
+	}
+	return sysInfo != nil, nil
+}
+
+// sysInfoIfExists returns the system info for the user's system. If the sysInfo
+// is missing, then nil is returned
+func (p *Package) sysInfoIfExists() (*lock.SystemInfo, error) {
 	if !featureflag.RemoveNixpkgs.Enabled() {
-		return false, nil
+		return nil, nil
 	}
 
 	if !p.isVersioned() {
-		return false, nil
+		return nil, nil
+	}
+
+	version, err := nix.Version()
+	if err != nil {
+		return nil, err
+	}
+
+	// enable for nix >= 2.17
+	if vercheck.SemverCompare(version, "2.17.0") < 0 {
+		return nil, err
 	}
 
 	entry, err := p.lockfile.Resolve(p.Raw)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
 
+	userSystem := nix.System()
+
 	if entry.Systems == nil {
-		return false, nil
+		return nil, nil
 	}
 
 	// Check if the user's system's info is present in the lockfile
-	_, ok := entry.Systems[nix.System()]
+	sysInfo, ok := entry.Systems[userSystem]
 	if !ok {
+		return nil, nil
+	}
+	return sysInfo, nil
+}
+
+// IsInBinaryCache returns true if the package is in the binary cache.
+// ALERT: Callers must call FillNarInfoCache before calling this function.
+func (p *Package) IsInBinaryCache() (bool, error) {
+
+	if eligible, err := p.isEligibleForBinaryCache(); err != nil {
+		return false, err
+	} else if !eligible {
 		return false, nil
 	}
 
-	version, err := nix.Version()
+	// Check if the narinfo is present in the binary cache
+	isNarInfoInCache.lock.RLock()
+	exists, ok := isNarInfoInCache.status[p.Raw]
+	isNarInfoInCache.lock.RUnlock()
+	if !ok {
+		return false, errors.Errorf("narInfo cache miss: %v. call XYZ before invoking IsInBinaryCache", p.Raw)
+	}
+	return exists, nil
+}
+
+// fillNarInfoCache fills the cache value for the narinfo of this package,
+// if it is eligible for the binary cache.
+func (p *Package) fillNarInfoCache() error {
+	if eligible, err := p.isEligibleForBinaryCache(); err != nil {
+		return err
+	} else if !eligible {
+		return nil
+	}
+
+	sysInfo, err := p.sysInfoIfExists()
```

Comment:
Gah, you are right.

Reply:
Thought: Instead of sprinkling locks everywhere, I could ensure these are invoked for all packages prior to starting the go-routines. Then within the goroutines, they'll be in read-only mode. Would that work?

Reply:
I think that would work. It might be easier to just split apart the HTTP request from everything else. Then you can calculate the store hash and check the map synchronously.

Reply:
nix.Version is cached already